### PR TITLE
Make detectOptional methods throw NullPointerException when detecting null value.

### PR DESCRIPTION
**Pull request by itohro**
Revisiting the comment below from #115.

>There's one weird case where it's not ok to delegate to ofNullable(). The map, or iterable, could contain a null value and the predicate could be each -> each == null. There's no great way to handle that case, but it's not great to return empty() when something was actually found. The alternative would be calling Optional.of(null) which throws, but is arguably more correct.

Following what's suggested in the last line above and also what `Streams#findFirst()` does, I made detectOptional methods throw NullPointerException when null value is detected.

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/CompositeIterable.java [hunk 1]
```diff
@@ -100,6 +101,17 @@ public E detect(Predicate<? super E> predicate)
         return null;
     }
 
+    @Override
+    public Optional<E> detectOptional(Predicate<? super E> predicate)
+    {
+        for (int i = 0; i < this.iterables.size(); i++)
+        {
+            Iterable<E> eachIterable = this.iterables.get(i);
+            return Iterate.detectOptional(eachIterable, predicate);
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/CompositeIterable.java]: This doesn't look correct to me. The loop should proceed if an intermediate result is empty. This and detect() are probably missing some edge case tests.

Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/CompositeIterable.java]: Good catch! I added missing test cases in `CompositeIterableTest` (and also in `AbstractLazyIterableTestCase`) and fixed this.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java [hunk 2]
```diff
@@ -78,6 +79,24 @@ public V detect(Predicate<? super V> predicate)
     }
 
     @Override
+    public Optional<V> detectOptional(Predicate<? super V> predicate)
+    {
+        V[] result = (V[]) new Object[1];
+        Iterate.detectOptional(this.adapted, each -> {
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: The inner call to detectOptional() here, and detect() in the previous method, is just used to short-circuit. The result is not stored in a variable. I think it would be more clear to use anySatisfy() in both methods, in a way similar to the inner-most call to anySatisfy(). The way it is written will work, but it makes it seem like the Optional returned by detectOptional() is significant.

Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: This call is actually logically significant in a way that it should throw NullPointerException when it detects the value which is null. Throwing NullPointerException is handled within Iterate.detectOptional(), when Optional.of() is called.

Having said that, we could change the detect() side as you suggest, but maybe keeping as is makes the structure consistent. Let me know.

Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: The call to detectOptional is on the elements of type T before they are
flattened to V. If the iterable itself is null, we'll get a NPE simply by
iterating over null. Is there something else accomplished by the call?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/ImmutableArrayList.java [hunk 3]
```diff
@@ -282,13 +282,29 @@ public T detect(Predicate<? super T> predicate)
...
 
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> function)
+    {
+        T result = this.detectWith(predicate, parameter);
+        return result == null ? function.value() : result;
+    }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/ImmutableArrayList.java]: The override for detectWith() looks fine because it delegates to InternalArrayIterate. I don't understand why detectWithIfNone() is improved by an override. And detectIfNone() is identical to the method it overrides, should we remove it?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java [hunk 4]
```diff
@@ -125,13 +126,41 @@ public T detect(Predicate<? super T> predicate)
...
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> function)
+    {
+        T result = this.detectWith(predicate, parameter);
+        return result == null ? function.value() : result;
+    }
+
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java]: detectIfNone() is identical to the method it overrides. detectWithIfNone is very slightly different.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java [hunk 5]
```diff
@@ -355,13 +356,41 @@ public T detect(Predicate<? super T> predicate)
...
 
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> function)
+    {
+        T result = ListIterate.detectWith(this, predicate, parameter);
+        return result == null ? function.value() : result;
+    }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: I don't see how this override of detectIfWithNone is better than the implementation it overrides. It goes through a different path but still winds up calling detect(). And it's one more place with a null check. Could we just skip the override of detectIfNone and detectWithIfNone?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java (hunk 2 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: Actually you are right. Here are how I fixed this.

1. I added missing test case in FlatCollectIterableTestNoIteratorTest.
2. I applied your suggestion to use anySatisfy() for both detect() and detectOptional()
3. Fixed the logic to explicitly throw NPE when iterable is null (otherwise it'll throw IllegalArgumentException)
4. Fixed the logic to explicitly throw NPE when detecting null value

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/ImmutableArrayList.java (hunk 3 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/immutable/ImmutableArrayList.java]: That's true. I removed these.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java (hunk 4 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/fixed/AbstractArrayAdapter.java]: These has been there before I modify (I just moved method location), but agree. I removed both as detectWithIfNone also became identical after addressing another comment in `AbstractMutableList`.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java (hunk 5 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/AbstractMutableList.java]: Agree and removed these overrides, too.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java [hunk 6]
```diff
@@ -78,6 +79,32 @@ public V detect(Predicate<? super V> predicate)
...
+                throw new NullPointerException();
+            }
+            return Iterate.anySatisfy(iterable, each1 -> {
+                if (predicate.accept(each1))
+                {
+                    if (each == null)
+                    {
+                        throw new NullPointerException();
+                    }
+                    result[0] = each1;
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: I don't understand checking `each == null` all the way down here. If we made it here, then `function.valueOf(each)` already returned a non-null, non-empty list. I think we should either check earlier in the outer lambda, or not at all.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java [hunk 7]
```diff
@@ -905,13 +905,41 @@ public T detect(Predicate<? super T> predicate)
...
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> defaultValueBlock)
+    {
+        T result = this.detectWith(predicate, parameter);
+        return result == null ? defaultValueBlock.value() : result;
+    }
+
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: This implementation of detectIfNone is identical to the superclass, and the implementation of both detechIfNone and detectWithIfNone have the null check for the result. An easy way to replace all the remaining places that check `result == null` would be something like this.

```java
return this.detectOptional(predicate).orElseGet(defaultValueBlock::value);
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/UnifiedSet.java [hunk 8]
```diff
@@ -773,6 +774,32 @@ else if (cur != null)
...
+                if (predicate.accept(each))
+                {
+                    return Optional.of(each);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
     private Object chainedDetect(ChainedBucket bucket, Predicate<? super T> predicate)
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/set/mutable/UnifiedSet.java]: Very nice, the return value of chainedDetect() is already unambiguous and only returns null to mean not found.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/immutable/sorted/ImmutableSortedBagImplTest.java [hunk 9]
```diff
@@ -16,10 +16,11 @@
 import org.eclipse.collections.impl.block.factory.Comparators;
 import org.eclipse.collections.impl.test.junit.Java8Runner;
 import org.eclipse.collections.test.IterableTestCase;
+import org.eclipse.collections.test.NoDetectOptionalNullTestCase;
 import org.junit.runner.RunWith;
 
 @RunWith(Java8Runner.class)
-public class ImmutableSortedBagImplTest implements ImmutableSortedBagTestCase
+public class ImmutableSortedBagImplTest implements ImmutableSortedBagTestCase, NoDetectOptionalNullTestCase
 {
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/immutable/sorted/ImmutableSortedBagImplTest.java]: The detectOptional null test irrelevant here because ImmutableSortedBag cannot contain nulls. ImmutableSortedBag cannot contain nulls because its elements are sorted with a Comparator, which doesn't tolerate null. Isn't this assertion irrelevant for all SortedIterables? If so, you can delete NoDetectOptionalNullTestCase and move its method into SortedIterableTestCase.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java (hunk 6 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: Good catch! 

I actually meant to check `each1` here. I reviewed my test case and find one corner case is missing there. I added missing test in `FlatCollectIterableTestNoIteratorTest`, moved `each` check out of lambda and put `each1` check here.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 7 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: Good suggestion. I only made this change in `AbstractRichIterable` for now. I'll do the whole refactoring in another PR.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/immutable/sorted/ImmutableSortedBagImplTest.java (hunk 9 above)
Comment (itohro) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/immutable/sorted/ImmutableSortedBagImplTest.java]: Good suggestion. I put this to SortedIterableTestCase and removed from its subclasses.

**Comment by motlin** [comment]
@itohro the build is currently failing on this PR.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 7 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: There is one interesting edge case where this doesn't work.

```java
Assert.assertEquals(Integer.valueOf(3), this.newWith(Comparators.reverseNaturalOrder(), 2, 3, 4, 5).detectIfNone(Integer.valueOf(3)::equals, null));
```

When it detects an element that meets the predicate condition, it should return it even if the function passed is null. However it throws NullPointerException with the method reference approach above (it seems method reference is "evaluated" even though it is not executed in the code path). 

We can change it as below to address this particular case still using detectOptional().

```java
this.detectOptional(predicate).orElseGet(() -> (function == null) ? null : function.value());
```

However now this code behaves differently from the original code when it detects null AND null function is passed in. Original code throws NPE, while the code above returns null.

I think this behaviour was not documented or tested anywhere, but I suppose original code behaves more naturally in this case (throws NPE). 

So I'd revert this to be original null check code. Let me know what you think.

Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: That doesn't sound correct to me. Did you make sure to pass `defaultValueBlock::value` instead of `defaultValueBlock`?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java [hunk 10]
```diff
@@ -78,6 +79,36 @@ public V detect(Predicate<? super V> predicate)
...
+                throw new NullPointerException();
+            }
+            return Iterate.anySatisfy(iterable, each1 -> {
+                if (predicate.accept(each1))
+                {
+                    if (each1 == null)
+                    {
+                        throw new NullPointerException();
+                    }
+                    result[0] = each1;
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: If you deleted this if-statement, wouldn't we get a NullPointerException on line 108 instead, from Optional.ofNullable(null)? Seems redundant?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 7 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: Yes, this was the code that was failing with NPE (it was in `AbstractRichIterable` not in `FastList` so using `function` instead of `defaultValueBlock`).

```
    @Override
    public T detectIfNone(Predicate<? super T> predicate, Function0<? extends T> function)
    {
        return this.detectOptional(predicate).orElseGet(function::value);
    }
```

Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: If detectOptional returns a present Optional, then function should not be
invoked and passing a null function should be ok.

Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: This is the simplest test case to simulate it.

```java
    @Test
    public void testNullMethodRefEval()
    {
        Function0<String> nullFunction = null;
        Verify.assertThrows(NullPointerException.class, () -> this.methodRef(nullFunction::value));
    }

    private String methodRef(Supplier<String> nullFunctionSupplier)
    {
        return "I don't execute the function but it throws NPE";
    }
```

Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: Fascinating. The method turns out to be irrelevant. This alone throws.
```java
Function0<String> nullFunction = null;
Supplier<String> supplier = nullFunction::value;
```

Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: And this doesn't throw.
```java
Function0<String> nullFunction = null;
Supplier<String> supplier = () -> nullFunction.value();
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java (hunk 10 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/lazy/FlatCollectIterable.java]: We have to differentiate the case we detect null (we only know it by checking `each1` == null after evaluating predicate.accept(each1)) vs the case we didn't find anything (result[0] stays null).

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java (hunk 7 above)
Comment (itohro) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java]: This is interesting :) 

I can do this lambda way, but again, we could do this in a separate PR as there are anyways couple of places we could apply this.
